experimental/sql/dialects/rel_impl.py

```diff
@@ -0,0 +1,354 @@
+# of operations: Expressions and Operators. Operators work on Bags whereas
+# expressions work on single values and are used to specify the exact behavior
+# of operators. The IndexByName and Yield Expressions bridge the gap to convert
+# from one to the other.
+
+#===------------------------------------------------------------------------===#
+# Data types
+#===------------------------------------------------------------------------===#
+
+
+@irdl_attr_definition
+class DataType(ParametrizedAttribute):
+  """
+  Models a datatype in a relational implementation query.
+  """
+  name = "rel_impl.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational implementation query.
+
+  Example:
+
+  ```
+  !rel_impl.int32
+  ```
+  """
+  name = "rel_impl.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational implementation query, that can be either
+  nullable or not.
+
+  Example:
+
+  ```
+  !rel_impl.string<0: !i1>
+  ```
+  """
+  name = "rel_impl.string"
+
+  # TODO: redefine nullable as a property of all fields
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: Union[int, IntegerAttr]) -> 'String':
+    if isinstance(val, IntegerAttr):
+      return String([val])
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+#===------------------------------------------------------------------------===#
+# Query types
+#===------------------------------------------------------------------------===#
+
+
+@irdl_attr_definition
+class Boolean(ParametrizedAttribute):
+  """
+  Models a type that can either be true or false to, e.g., show whether a tuple
+  is part of the result or not.
+
+  Example:
+
+  '''
+  !rel_impl.bool
+  '''
+  """
+  name = "rel_impl.bool"
+
+
+@irdl_attr_definition
+class SchemaElement(ParametrizedAttribute):
+  """
+  Models an element of a schema with name `elt_name` and type `elt_type`.
+
+  Example:
+  '''
+  !rel_impl.schema_element<"id", !rel_impl.int32>
+  '''
+  """
+  name = "rel_impl.schema_element"
+
+  elt_name = ParameterDef(StringAttr)
+  elt_type = ParameterDef(DataType)
+
+  @staticmethod
+  @builder
+  def get(name: str, type_: DataType) -> 'SchemaElement':
+    return SchemaElement([StringAttr.from_str(name), type_])
+
+
+@irdl_attr_definition
+class Bag(ParametrizedAttribute):
+  """
+  Models a bag in a relational implementation query. The exact schema of the bag
+  is part of the type itself.
+
+  Example:
+
+  '''
+  !rel_impl.bag<[!rel_impl.schema_element<"id", !rel_impl.int32>]>
+  '''
+  """
+  name = "rel_impl.bag"
+
+  schema = ParameterDef(ArrayOfConstraint(SchemaElement))
+
+  @staticmethod
+  @builder
+  def get(types: list[DataType], names: list[str]) -> 'Bag':
+    schema_elts = [SchemaElement.get(n, t) for n, t in zip(names, types)]
+    return Bag([ArrayAttr.from_list(schema_elts)])
+
+
+@irdl_attr_definition
+class Tuple(ParametrizedAttribute):
```

Comment:
As far as I can tell, this is the only difference from the previous dialect, right?

I totally support the strategy to start with the "maximum" number of dialects and then reduce but this seems like a good candidate to merge... 😉

Reply:
I 100% agree here. I am not quite sure which version we would take in the merger, probably this lower-level one because it is probably simpler to lower.

The main thing I can think of that will be added as well is the implementations of joins, which is not a huge change either :)